RHOAIENG-25116: Adding guidelines for KEDA autoscaling (#937)
opendatahub-io/opendatahub-documentation: modules/guidelines-for-metrics-based-autoscaling.adoc

You can control how the Horizontal Pod Autoscaler (HPA) scales down your application by managing the KEDA `scaledObject` resource. This resource is a Custom Resource (CR) that enables event-driven autoscaling for a specific workload.

To configure the scale-down behavior of the HPA, you must adjust the `advanced.horizontalPodAutoscalerConfig.behavior.scaleDown.stabilizationWindowSeconds` field within the `scaledObject` CR as shown in the example:

[source, YAML]
----
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: my-app-scaler
spec:
  advanced:
    horizontalPodAutoscalerConfig:
      behavior:
        scaleDown:
          stabilizationWindowSeconds: 300

----
The optimal configuration depends on finding the right balance between resource efficiency and application responsiveness for your particular workload requirements.

> == Considering model size for optimal scaling

Reviewer: ```suggestion
== Selecting a model size for optimal scaling
```

Reviewer 2: This changes the original meaning.

Author: @eturner24 Will keep this title as-is, because we're not selecting a model size here.

> When implementing metrics-based autoscaling for AI inference workloads, you must consider the impact of model size on scaling behavior and resource allocation strategies. The specifications below outline default configurations and typical performance characteristics to guide your approach.

Reviewer: ```suggestion
Model size affects autoscaling behavior and resource use. The following table describes the typical characteristics for different model sizes and describes a scaling strategy to select when implementing metrics-based autoscaling for AI inference workloads.
```

> * **Small models (< 3B parameters):**
>     ** Memory footprint: < 6 GiB.
>     ** Suitable for aggressive scaling policies with lower resource buffers.
>     ** Cold start times: Up to 10 minutes to download and 30 seconds to load.
> * **Medium models (3B-10B parameters):**
>     ** Memory footprint: 6-20 GiB.
>     ** Require more conservative scaling due to the size of the model.
>     ** Cold start times: Up to 30 minutes to download and 1 minute to load.
> * **Large models (> 10B parameters):**
>     **  Memory footprint: > 20 GiB.
>     **  May require model sharding or quantization for efficient scaling.
>     **  Cold start times: Up to hours to download and several minutes to load.

Reviewer: ```suggestion
[cols="1,1,1,1", options="header"]  
|=== 
| Model size | Memory footprint | Scaling strategy | Cold start time
| Small (Less than 3B)
| Less than 6 GiB
| Use aggressive scaling with lower resource buffers.
| Up to 10 minutes to download and 30 seconds to load.
| Medium (3B-10B)
| 6-20 GiB	
| Use a more conservative scaling strategy.	
| Up to 30 minutes to download and 1 minute to load.
| Large (Greater than 10B)
| Greater than 20 GiB
| May require model sharding or quantization.
| Up to several hours to download and minutes to load.
|===
```

> For models with fewer than 3 billion parameters, you can use various techniques to significantly reduce cold start latency:

Reviewer: ```suggestion
For models with fewer than 3 billion parameters, you can reduce cold start latency with the following strategies:
```
